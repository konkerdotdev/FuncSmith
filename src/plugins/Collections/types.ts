import type { FileSetItem } from '../../lib/fileSet';

export type Collection<T extends FileSetItem> = T & {
  readonly collection: {
    readonly len: number;
    readonly previous?: number;
    readonly next?: number;
  };
};

export type CollectionOptions = {
  readonly globPattern: string;
  readonly sortBy: string;
  readonly reverse: boolean;
};

export type Convenience<T extends Partial<CollectionOptions>> = string | T;
