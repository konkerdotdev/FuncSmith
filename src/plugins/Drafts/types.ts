export type DraftsOptions = {
  readonly default: boolean;
  readonly include: boolean;
};
