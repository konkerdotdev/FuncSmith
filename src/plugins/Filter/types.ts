export type FilterOptions = {
  readonly drop: Array<string>;
  readonly keep: Array<string>;
};
