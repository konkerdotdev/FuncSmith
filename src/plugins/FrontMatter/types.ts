import type { FileSetItem } from '../../lib/fileSet';

export type FrontMatter<T extends Partial<FileSetItem>> = T & {
  readonly frontMatter: Record<string, unknown>;
};

export type FrontMatterOptions = {
  readonly globPattern: string;
};
