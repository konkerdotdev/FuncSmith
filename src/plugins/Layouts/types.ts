import type * as H from 'handlebars';

export type LayoutsOptions = {
  readonly templateEngine: string;
  readonly directory: string;
  readonly defaultLayout: string;
  readonly globPattern: string;
  readonly helpers: Record<string, H.HelperDelegate>;
};
