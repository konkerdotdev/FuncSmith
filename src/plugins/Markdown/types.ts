export type MarkdownOptions = {
  readonly globPattern: string;
};
