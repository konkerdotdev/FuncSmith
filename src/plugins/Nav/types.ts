import type { FileSetItem } from '../../lib/fileSet';

export type NavItem<T extends FileSetItem> = T & {
  readonly nav: boolean;
};

export type NavOptions = {
  readonly navProperty: string;
  readonly sortBy: string;
  readonly reverse: boolean;
};
