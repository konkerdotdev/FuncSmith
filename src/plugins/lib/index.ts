import * as P from '@konker.dev/effect-ts-prelude';

import type { FileSet, FileSetItem } from '../../lib/fileSet';
import type { FileSetInjectionMappingCtor, FileSetMapping, FileSetMappingCtor } from '../../types';

/**
 * A type when the plugin wants to accept either a full options record, or a string convenience version.
 * E.g.
 * ```
 *     F.collections({
 *       posts: {
 *         globPattern: 'posts/*.html',
 *         sortBy: 'date',
 *       },
 *     }),
 * ```
 * vs
 * ```
 *     F.collections({
 *       posts: 'posts/*.html',
 *     }),
 * ```
 */
export type Convenience<T> = string | Partial<T>;

/**
 * IF - input fileSetItem
 * OF - output fileSetItem
 * MF - mapping output fileSetItem
 * RM - mapping deps
 * RN - next mapping deps
 * OP - options
 */
export const wrapMapping =
  <IF extends FileSetItem, MF extends FileSetItem, RM, OP = undefined>(
    mappingCtor: FileSetMappingCtor<IF, MF, RM, OP>
  ) =>
  (options?: OP) =>
  <OF extends FileSetItem, RN>(next: FileSetMapping<MF, OF, RN>) =>
  (fileSet: FileSet<IF>) =>
    P.pipe(P.Effect.succeed(fileSet), P.Effect.flatMap(mappingCtor(options)), P.Effect.flatMap(next));

/**
 * IF - input fileSetItem
 * OF - output fileSetItem
 * MF - mapping output fileSetItem
 * RM - mapping deps
 * RN - next mapping deps
 * OP - options
 */
export const wrapInjection =
  <IF extends FileSetItem, OF extends FileSetItem, RN, RM, OP = undefined>(
    postMappingCtor: FileSetInjectionMappingCtor<OF, RN, RM, OP>
  ) =>
  (options: OP) =>
  (next: FileSetMapping<IF, OF, RN>): FileSetMapping<IF, OF, RM> =>
  (fileSet: FileSet<IF>) =>
    P.pipe(next(fileSet), postMappingCtor(options));
