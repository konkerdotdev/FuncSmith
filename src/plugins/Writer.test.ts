import * as P from '@konker.dev/effect-ts-prelude';
import { fs } from 'memfs';
import { toTreeSync } from 'memfs/lib/print';

import { FuncSmithContextSinkTest, FuncSmithContextWriterTest } from '../layers';
import * as fixturesFsFm from '../test/fixtures/fileset-frontmatter-1';
import { FuncSmithContextSink } from '../types';
import * as unit from './Writer';

describe('plugins', () => {
  describe('Writer', () => {
    it('should return the same value', async () => {
      const pluginStack = P.pipe(P.Effect.succeed, unit.writer());
      const actual = await P.Effect.runPromise(
        P.pipe(
          [...fixturesFsFm.TEST_FILE_SET_FRONT_MATTER_1],
          pluginStack,
          P.Effect.provideService(FuncSmithContextSink, FuncSmithContextSinkTest),
          P.Effect.provide(FuncSmithContextWriterTest)
        )
      );
      expect(actual).toStrictEqual(fixturesFsFm.TEST_FILE_SET_FRONT_MATTER_1);

      // Inspect state of memfs
      expect(toTreeSync(fs)).toMatchSnapshot('plugins-writer-1');
    });
  });
});
