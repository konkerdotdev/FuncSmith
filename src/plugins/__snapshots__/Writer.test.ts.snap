// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins Writer should return the same value: plugins-writer-1 1`] = `
"/
└─ tmp/
   ├─ build/
   │  ├─ a.txt
   │  ├─ docs/
   │  │  ├─ d1.doc
   │  │  ├─ d2.doc
   │  │  └─ d3.doc
   │  └─ posts/
   │     ├─ p1.md
   │     ├─ p2.md
   │     ├─ p3.md
   │     └─ p4.md
   ├─ foo/
   │  ├─ a.txt
   │  ├─ b.txt
   │  ├─ bar/
   │  │  ├─ e.txt
   │  │  └─ f.log
   │  ├─ c.csv
   │  ├─ d.json
   │  ├─ ignore1.txt
   │  └─ posts/
   │     ├─ index.md
   │     ├─ p1.md
   │     ├─ p2.md
   │     └─ p3.md
   ├─ layouts/
   │  ├─ bar.hbs
   │  ├─ foo.hbs
   │  ├─ layout.hbs
   │  └─ post.hbs
   └─ partials/
      └─ footer.hbs"
`;
